### PR TITLE
Add deadline functionality to Tasks

### PR DESCRIPTION
**Pull request by seantehds**

### CODE REVIEW

File: src/main/java/seedu/address/logic/commands/MarkCommand.java [hunk 1]
```diff
@@ -48,9 +48,13 @@ public CommandResult execute(Model model) throws CommandException {
...
 
-        Task markedTask = new Task(taskToMark.getDescription(), status.updateStatus());
-        model.setTask(taskToMark, markedTask);
-
+        if (taskToMark.getDeadline() == null) {
+            Task markedTask = new Task(taskToMark.getDescription(), status.updateStatus());
+            model.setTask(taskToMark, markedTask);
+        } else {
+            Task markedTask = new Task(taskToMark.getDescription(), status.updateStatus(), taskToMark.getDeadline());
+            model.setTask(taskToMark, markedTask);
```
Comment (asdfghjkxd) [inline src/main/java/seedu/address/logic/commands/MarkCommand.java:56]: Should we have a `NullDeadline` object that extends from `Deadline` to represent no deadline assigned to the Task, rather than just using `null`?

File: src/main/java/seedu/address/logic/commands/UnmarkCommand.java [hunk 2]
```diff
@@ -48,8 +48,14 @@ public CommandResult execute(Model model) throws CommandException {
...
 
-        Task unmarkedTask = new Task(taskToUnmark.getDescription(), status.updateStatus());
-        model.setTask(taskToUnmark, unmarkedTask);
+        if (taskToUnmark.getDeadline() == null) {
+            Task markedTask = new Task(taskToUnmark.getDescription(), status.updateStatus());
+            model.setTask(taskToUnmark, markedTask);
+        } else {
+            Task markedTask = new Task(taskToUnmark.getDescription(), status.updateStatus(),
+                    taskToUnmark.getDeadline());
+            model.setTask(taskToUnmark, markedTask);
```
Comment (asdfghjkxd) [inline src/main/java/seedu/address/logic/commands/UnmarkCommand.java:57]: Same comment as in `MarkCommand.java`.

File: src/main/java/seedu/address/logic/parser/ParserUtil.java [hunk 3]
```diff
@@ -49,6 +53,38 @@ public static Description parseDescription(String description) throws IllegalArg
...
+     * @throws IllegalArgumentException if the given {@code name} is invalid.
+     */
+    public static Deadline parseDeadline(String deadline) throws IllegalArgumentException {
+        requireNonNull(deadline);
+
+        deadline = deadline.strip();
+        if (Deadline.isValidDateTime(deadline)) {
+            //@@author asdfghjkxd-reused
+            // Regex strings are reused with major modification from ChatGPT, and is built and tested with
+            // https://regex101.com/.
```
Comment (asdfghjkxd) [inline src/main/java/seedu/address/logic/parser/ParserUtil.java:69]: Should this be replaced by @seantehds 's username instead?

File: src/main/java/seedu/address/model/task/Deadline.java [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+ */
+public class Deadline {
+
+    public static final String MESSAGE_CONSTRAINTS =
+            "Deadlines need to be in the format (DD(- OR /)MM(- OR /)YYYY HH(: or -)MM) "
+                    + "OR (DD(- OR /)MM(- OR /)YYYY).";
+
+    //@@author asdfghjkxd-reused
+    // Regex strings are reused with some modification from ChatGPT, and is also built and tested with
+    // https://regex101.com/.
```
Comment (asdfghjkxd) [inline src/main/java/seedu/address/model/task/Deadline.java:20]: Same with `ParserUtil.java`.

File: src/main/java/seedu/address/model/task/Task.java [hunk 5]
```diff
@@ -44,6 +47,21 @@ public Task(Description description, Status status) {
         this.description = description;
         this.status = status;
         this.tags.addAll(Collections.emptySet());
+        this.deadline = null;
```
Comment (asdfghjkxd) [inline src/main/java/seedu/address/model/task/Task.java:50]: Should we not use `null` directly over here?

File: src/main/java/seedu/address/ui/TaskCard.java [hunk 6]
```diff
@@ -46,6 +48,11 @@ public TaskCard(Task task, int displayedIndex) {
         id.setText(displayedIndex + ". ");
         description.setText(task.getDescription().fullDescription);
         status.setText(task.getStatus().toString());
+        if (task.getDeadline() == null) {
+            deadline.setText("");
```
Comment (asdfghjkxd) [inline src/main/java/seedu/address/ui/TaskCard.java:52]: Same as above, regarding the use of `null` as a valid value for Deadline.

**Review by asdfghjkxd — COMMENTED** [review]
Looks good! There are some nits that may need to be looked at (regarding the use of `null` as valid values and the author's name in code reuse).

File: src/test/java/seedu/address/logic/parser/ParserUtilTest.java [hunk 7]
```diff
@@ -28,6 +31,19 @@ public class ParserUtilTest {
     private static final String VALID_TAG_2 = "neighbour";
 
     private static final String WHITESPACE = " \t\r\n";
+    private static final String VALID_DEADLINE_WITH_TIME = "25-10-2023 16:00";
+    private static final String VALID_DEADLINE_WITHOUT_TIME = "25-10-2023";
```
Comment (JCSnap) [inline src/test/java/seedu/address/logic/parser/ParserUtilTest.java:35]: Just a question: Is it supposed to be intelligent enough to corroborate with real life data, or is it purely formatting?
For example, suppose I put the date for 29th of February on a non leap year, with the correct format, is it supposed to throw exceptions?

Comment (asdfghjkxd) [reply src/test/java/seedu/address/logic/parser/ParserUtilTest.java:35]: I believe that the internal LocalDataTime object used will check for the validity of the input datetime, but we could add examples of edge cases that are invalid dates to the test cases to test if the parsing actually works.

**Review by JCSnap — APPROVED** [review]
